File: slack-reader/src/test/java/dev/number6/slackreader/core/SlackServiceTest.java
package dev.number6.slackreader.core;

import dev.number6.slackreader.SlackService;
import dev.number6.slackreader.model.Channel;
import dev.number6.slackreader.model.Message;
import dev.number6.slackreader.model.WorkspaceMessages;
import dev.number6.slackreader.port.SlackReaderConfigurationPort;
import dev.number6.slackreader.port.SlackPort;
import dev.number6.slackreader.generate.SlackReaderRDG;
import com.amazonaws.services.lambda.runtime.LambdaLogger;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;
import org.mockito.Mock;
import org.mockito.junit.jupiter.MockitoExtension;
import uk.org.fyodor.generators.Generator;
import uk.org.fyodor.range.Range;

import java.time.LocalDate;
import java.util.*;
import java.util.stream.Collectors;

import static org.assertj.core.api.Assertions.assertThat;
import static org.mockito.ArgumentMatchers.any;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.when;

@ExtendWith(MockitoExtension.class)
class SlackServiceTest {

    @Mock
    private LambdaLogger logger;
    @Mock
    private SlackPort slackPort;
    @Mock
    private SlackReaderConfigurationPort config;

    private final Generator<Set<Channel>> channelsGenerator = SlackReaderRDG.channels(Range.closed(5, 20));
    private final Generator<List<Message>> messagesGenerator = SlackReaderRDG.list(SlackReaderRDG.message(), Range.closed(5, 50));
    private SlackService testee;

    @BeforeEach
    void setup() {
        testee = new SlackService(slackPort, config);
    }

    @Test
    void getMessagesOnDate() {

        Set<Channel> channels = channelsGenerator.next();
        Map<String, List<Message>> expectedMessages = channels.stream().collect(Collectors.toMap(Channel::getName, c -> messagesGenerator.next()));

        when(slackPort.getChannelList(logger)).thenReturn(channels);
        when(slackPort.getMessagesForChannelOnDate(any(), any(), any())).then(i -> expectedMessages.get(((Channel) i.getArgument(0)).getName()));

        WorkspaceMessages messages = testee.getMessagesOnDate(LocalDate.now(), logger);

        List<String> channelNames = channels.stream().map(Channel::getName).collect(Collectors.toList());
        assertThat(messages.getChannelNames()).containsExactlyInAnyOrderElementsOf(channelNames);

        messages.getChannelNames().forEach(name ->
                assertThat(messages.getMessagesForChannel(name))
                        .containsExactlyInAnyOrderElementsOf(expectedMessages.get(name).stream().map(Message::getText).collect(Collectors.toList())));
    }

    @Test
    void ignoreMessagesForBlacklistedChannels() {

        String channel1Name = "channel 1 name";
        String channel2Name = "channel 2 name";
        when(config.getBlacklistedChannels()).thenReturn(Arrays.asList(channel1Name, channel2Name));
        Set<Channel> channels = channelsGenerator.next();
        channels.add(new Channel("whatever", channel1Name));
        channels.add(new Channel("whatever2", channel2Name));
        when(slackPort.getChannelList(logger)).thenReturn(channels);

        WorkspaceMessages messages = testee.getMessagesOnDate(LocalDate.now(), logger);

        assertThat(messages.getChannelNames()).doesNotContain(channel1Name, channel2Name);
    }

}

File: slack-reader/src/test/java/dev/number6/slackreader/core/SlackServiceTest.kt
package dev.number6.slackreader.core

import assertk.assertThat
import assertk.assertions.containsNone
import assertk.assertions.containsOnly
import com.amazonaws.services.lambda.runtime.LambdaLogger
import dev.number6.slackreader.SlackService
import dev.number6.slackreader.generate.SlackReaderRDG
import dev.number6.slackreader.model.Channel
import dev.number6.slackreader.model.Message
import dev.number6.slackreader.port.SlackPort
import dev.number6.slackreader.port.SlackReaderConfigurationPort
import io.mockk.every
import io.mockk.junit5.MockKExtension
import io.mockk.mockk
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.extension.ExtendWith
import uk.org.fyodor.generators.Generator
import uk.org.fyodor.generators.RDG
import uk.org.fyodor.range.Range
import java.time.LocalDate

@ExtendWith(MockKExtension::class)
internal class SlackServiceTest {
    private val logger: LambdaLogger = mockk(relaxUnitFun = true)

    private val slackPort: SlackPort = mockk(relaxUnitFun = true)

    private val config: SlackReaderConfigurationPort = mockk(relaxUnitFun = true)

    private val channelsGenerator = SlackReaderRDG.channels(Range.closed(5, 20))
    private val messagesGenerator: Generator<List<Message>> = RDG.list(SlackReaderRDG.message(), Range.closed(5, 50))
    private val testee: SlackService = SlackService(slackPort, config)

    @Test
    fun getMessagesOnDate() {
        every { config.getBlacklistedChannels() } returns listOf()
        val channels = channelsGenerator.next()
        val expectedMessages = channels.associateBy({ c -> c.name }, { messagesGenerator.next() })
        every { slackPort.getChannelList(logger) } returns channels
        every { slackPort.getMessagesForChannelOnDate(any(), any(), any()) } answers {
            expectedMessages[firstArg<Channel>().name] ?: listOf()
        }
        val messages = testee.getMessagesOnDate(LocalDate.now(), logger)
        val channelNames = channels.map { c -> c.name }
        assertThat(messages.getActiveChannelNames()).containsOnly(*channelNames.toTypedArray())

        messages.getActiveChannelNames().forEach { name ->
            assertThat(messages.getMessagesForChannel(name))
                    .containsOnly(*expectedMessages[name]?.map { m -> m.text }?.toTypedArray() ?: Array(0) { "" })
        }
    }

    @Test
    fun ignoreMessagesForBlacklistedChannels() {
        val channel1Name = "channel 1 name"
        val channel2Name = "channel 2 name"
        every { config.getBlacklistedChannels() } returns listOf(channel1Name, channel2Name)
        val channels = channelsGenerator.next()
        channels.add(Channel("whatever", channel1Name))
        channels.add(Channel("whatever2", channel2Name))
        every { slackPort.getChannelList(logger) } returns channels
        every { slackPort.getMessagesForChannelOnDate(any(), any(), any()) } returns messagesGenerator.next()
        val messages = testee.getMessagesOnDate(LocalDate.now(), logger)
        assertThat(messages.getActiveChannelNames()).containsNone(channel1Name, channel2Name)
    }
}